### ISSUE TITLE
Failed to Index Large File

### ISSUE THREAD

**Issue by Bojun-Feng** [run-llama/LlamaIndexTS#483]
I have created a RAG app with `create-llama` and updated the data with the following large files:

[Fannie Mae Selling Guide.txt](https://github.com/run-llama/LlamaIndexTS/files/14095074/Fannie.Mae.Selling.Guide.txt)
[Freddie Mac Selling Guide.txt](https://github.com/run-llama/LlamaIndexTS/files/14095077/Freddie.Mac.Selling.Guide.txt)
[FHA Guidelines.txt](https://github.com/run-llama/LlamaIndexTS/files/14095075/FHA.Guidelines.txt)

(I understand that indexing files this large is burning tokens and is probably a really stupid idea. I just wanted to do a test run to see how the code would work under large files.)

Here is the terminal output:
```
$ npm run generate

> project@0.1.0 generate
> node app/api/chat/engine/generate.mjs

Generating storage context...
file:///workspaces/project/node_modules/openai/error.mjs:40
            return new BadRequestError(status, error, message, headers);
                   ^

BadRequestError: 400 This model's maximum context length is 8192 tokens, however you requested 20142 tokens (20142 in your prompt; 0 for the completion). Please reduce your prompt; or completion length.

Node.js v20.11.0
```

From the message, it seems like OpenAI is complaining about the prompt being too long. However, that doesn't make a lot of sense as I expected Llama Index to keep track of the chunk size and prevent the prompt from overflowing. Did this happen because I messed up, or is there a bug that needs to be fixed?

## How to replicate

1. Run `npx create-llama@latest` using streaming, Next.js, and local file structure
2. Put the above three files into the data folder
3. Run `npm run generate`

**Comment by marcusschiesser** [comment]
@Bojun-Feng, thanks for the report, currently indexing large files is very inefficient as we haven't added batching yet - it's coming soon with https://github.com/run-llama/LlamaIndexTS/pull/407 

Meanwhile, you can try local embedding in your `generate.mjs` script:
```
 const embedModel = new HuggingFaceEmbedding({
    modelType: HuggingFaceEmbeddingModelType.XENOVA_ALL_MPNET_BASE_V2,
  });
  const serviceContext = serviceContextFromDefaults({
     chunkSize: CHUNK_SIZE,
    chunkOverlap: CHUNK_OVERLAP,
    embedModel,
  });
```

**Comment by marcusschiesser** [comment]
needs to be retested after https://github.com/run-llama/LlamaIndexTS/pull/524

**Comment by revolunet** [comment]
Getting same issue v0.1.12 and OpenAI, using SimpleDirectoryReader

**Comment by marcusschiesser** [comment]
I just retried this test case with the latest llamaindex release (sending embeddings in batches) and I can reproduce the issue.

@revolunet Thanks for your help - I guess this means our text splitter is producing too long strings for OpenAI's embedding model - we're having a look into this

**Comment by marcusschiesser** [comment]
Is also reproducible with just using `FHA.Guidelines.txt`

Next step:
1. copy https://github.com/run-llama/LlamaIndexTS/blob/main/packages/create-llama/templates/components/vectordbs/typescript/none/generate.mjs to /examples (no create-llama needed)
2. debug `VectorStoreIndex.fromDocuments` (will calc the embeddings)

**Comment by marcusschiesser** [comment]
Just released llamaindex@0.1.18, which adds a `splitLongSentences` option to the `SimpleNodeParser` - usually, sentences are not split (which is better for retrieval), but the test document contains at the end an index that doesn't contain a sentence limiter. 

To use `splitLongSentences` have a look at this example:

```
const serviceContext = serviceContextFromDefaults({
    nodeParser: new SimpleNodeParser({
      chunkSize: 512,
      chunkOverlap: 20,
      splitLongSentences: true,
    }),
  });

  // generate a document with a very long sentence (9000 words long)
  const longSentence = "is ".repeat(9000) + ".";
  const document = new Document({ text: longSentence, id_: "1" });
  await VectorStoreIndex.fromDocuments([document], {
    serviceContext,
  });
```

@Bojun-Feng please let me know if that works for you, then we can close this ticket.

**Comment by marcusschiesser** [comment]
@Bojun-Feng closing this for now, if it doesn't work, please reopen
